cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java

```diff
@@ -49,8 +52,10 @@
 
   private static final String DATASET_CLASS_NAME = Dataset.class.getName();
   private final DatasetClassRewriter datasetRewriter;
+  private final AuthEnforceRewriter authEnforceRewriter;
   private final Function<String, URL> resourceLookup;
   private final Map<String, Boolean> cache;
+  private RewritesNeeded rewritesNeeded;
```

Comment:
Maybe better just call `Classes.isSubTypeOf` again in the `rewriteClass` method (the decision is cached in the `cache` map, so the overhead is super small, also it's easy to just turn the map to concurrent map if we decided to enable parallel classloading).